src/shared_realm.cpp

```diff
@@ -161,7 +161,9 @@ void Realm::open_with_config(const Config& config,
             };
             shared_group = std::make_unique<SharedGroup>(*history, options);
 
-            if (realm && config.should_compact_on_launch_function) {
+            auto coordinator = RealmCoordinator::get_existing_coordinator(config.path);
+
+            if (realm && config.should_compact_on_launch_function && !coordinator->has_initialized_realms()) {
```

Comment:
I think this block makes more sense in Realm's constructor than here. We don't want to do the compact when we're calling `open_with_config()` from anywhere but the constructor (it's used for things like opening the helper SGs for the notification worker), and there we already have a pointer to the coordinator.